components/layout/display_list_builder.rs

```diff
@@ -1738,51 +1738,89 @@ impl BlockFlowDisplayListBuilding for BlockFlow {
         // not recurse on a line if nothing in it can intersect the dirty region.
         debug!("Flow: building display list for {} inline fragments", self.fragments.len());
 
         let mut display_list = box DisplayList::new();
-        let mut has_stacking_context = false;
-        for fragment in &mut self.fragments.fragments {
-            fragment.build_display_list(&mut *display_list,
-                                        layout_context,
-                                        &self.base.stacking_relative_position,
-                                        &self.base
-                                             .early_absolute_position_info
-                                             .relative_containing_block_size,
-                                        self.base
-                                            .early_absolute_position_info
-                                            .relative_containing_block_mode,
-                                        BorderPaintingMode::Separate,
-                                        BackgroundAndBorderLevel::Content,
-                                        &self.base.clip,
-                                        &self.base.stacking_relative_position_of_display_port);
-
-            has_stacking_context = fragment.establishes_stacking_context();
-
-            match fragment.specific {
-                SpecificFragmentInfo::InlineBlock(ref mut block_flow) => {
-                    let block_flow = flow_ref::deref_mut(&mut block_flow.flow_ref);
-                    display_list.append_from(
-                        &mut flow::mut_base(block_flow).display_list_building_result)
-                }
-                SpecificFragmentInfo::InlineAbsoluteHypothetical(ref mut block_flow) => {
-                    let block_flow = flow_ref::deref_mut(&mut block_flow.flow_ref);
-                    display_list.append_from(
-                        &mut flow::mut_base(block_flow).display_list_building_result)
-                }
-                SpecificFragmentInfo::InlineAbsolute(ref mut block_flow) => {
-                    let block_flow = flow_ref::deref_mut(&mut block_flow.flow_ref);
-                    display_list.append_from(
-                        &mut flow::mut_base(block_flow).display_list_building_result)
+
+        // We iterate using an index here, because we want to avoid doing a doing
+        // a double-borrow of self (one mutable for the method call and one immutable
+        // for the self.fragments.fragment iterator itself).
+        for index in (0..self.fragments.fragments.len()) {
+            let establishes_stacking_context = {
+                let fragment = self.fragments.fragments.get(index).unwrap();
+                fragment.establishes_stacking_context() && match fragment.specific {
+                    SpecificFragmentInfo::InlineBlock(_) |
+                        SpecificFragmentInfo::InlineAbsoluteHypothetical(_) => false,
+                    _ => true,
```

Comment:
Instead of `true`, how about just `fragment.establishes_stacking_context()` (and then remove the `&&` and stuff before it)?